scripts/zephyr_module.py

```diff
@@ -50,6 +50,22 @@
         type: seq
         sequence:
           - type: str
+      settings:
+        required: false
+        type: map
+        mapping:
+          board_root:
+            required: false
+            type: str
```

Comment:
Nice, I like this approach.

Looks like these are relative paths from the root of the module which are made absolute below in process_settings().

Maybe it would be a good time to discuss https://github.com/zephyrproject-rtos/zephyr/issues/23825 in this context. What should a relative path mean when set by the user? Relative to the application directory, still? Will the various ROOT support sections be able to handle that with this change? Not in this PR, just in general.

Reply:
Agree, #23825 is worth a good discussion, but let's try not to go to deep with that in this PR.
As a Zephyr module can be located anywhere, and relocated at will, then the only thing we know inside the module is relative to its own root.

But for a user providing a `-DBOARD_ROOT=<relative-path>` then I would say that should be relative to his current working dir.
So if he is in:
```
cd /tmp
cmake ... -Dx_ROOT=<relative-path> -Bbuild /path/to/app
```
then path is relative to `/tmp`, and if he does:
```
cd /path/to/app
cmake ... -Dx_ROOT=<relative-path> -Bbuild .
```
then relative to `/path/to/app`

> Will the various ROOT support sections be able to handle that with this change? Not in this PR, just in general.

In this PR, and in relation to Zephyr modules, it will work.
In relation to `-Dx_ROOT=<relative-path>`, then I would need to dig deeper into how that is currently handled.

Reply:
>  As a Zephyr module can be located anywhere, and relocated at will, then the only thing we know inside the module is relative to its own root.

Yep, I agree that is the right approach in this particular case. I just wanted to think about whether we're taking any steps in this PR that would preclude solving #23825.

I see now that we can do an `if(IS_ABSOLUTE ...)` check later on, so I think the answer to that question is "no". We can treat relative paths in these lists as undefined behavior for now, and define proper behavior for them later.

> Agree, #23825 is worth a good discussion, but let's try not to go to deep with that in this PR.

What do you think about at least issuing a CMake warning `if(NOT IS_ABSOLUTE ${some_root})` in the loops that go over these ROOT variables?

Since we are making the behavior of these variables more complex here and the user-visible portions are starting to use relative paths, it might be nice to give the user some hints if they are doing something "by hand" that we know won't work.

Reply:
> > As a Zephyr module can be located anywhere, and relocated at will, then the only thing we know inside the module is relative to its own root.
> 
> Yep, I agree that is the right approach in this particular case. I just wanted to think about whether we're taking any steps in this PR that would preclude solving #23825.

No, we're not.
The `zephyr_module.py` will always provide full path, and using `set(DTS_ROOT <path>)` or using `-DDTS_ROOT=<path>` does not impact that.

#23825 is an independent issue.

> 
> I see now that we can do an `if(IS_ABSOLUTE ...)` check later on, so I think the answer to that question is "no". We can treat relative paths in these lists as undefined behavior for now, and define proper behavior for them later.
> 
> > Agree, #23825 is worth a good discussion, but let's try not to go to deep with that in this PR.
> 
> What do you think about at least issuing a CMake warning `if(NOT IS_ABSOLUTE ${some_root})` in the loops that go over these ROOT variables?
> 
> Since we are making the behavior of these variables more complex here and the user-visible portions are starting to use relative paths, it might be nice to give the user some hints if they are doing something "by hand" that we know won't work.

I think there is a big difference in using `-Dx_ROOT=<relative-path>` and `set(DTS_ROOT <relative-path>)` compared to a Zephyr module.
The Zephyr modules are always (and has always) been using relative path from their module root.
This goes for the new build settings, but also for the existing CMake and Kconfig settings.
Then `zephyr_module.py` ensures to expand those settings into absolute path for CMake and Kconfig.

Whereas #23825 only covers relative path in relation to CMake.